crates/red_knot_python_semantic/src/semantic_index/builder/except_handlers.rs

```diff
@@ -0,0 +1,128 @@
+            .last()
+            .expect("There should always be at least one `TryBlockContexts` on the stack")
+    }
+
+    /// Pop a new [`TryNodeContextStack`] off the stack of stacks.
+    ///
+    /// Each [`TryNodeContextStack`] is only valid for a single scope
+    pub(super) fn exit_scope(&mut self) {
+        let popped_context = self.0.pop();
+        assert!(
+            popped_context.is_some(),
+            "pop_context() should never be called on an empty stack \
+(this indicates an unbalanced `push_context()`/`pop_context()` pair of calls)"
+        );
+    }
+}
+
+/// The contexts of nested `try`/`except` blocks for a single scope
+#[derive(Debug, Default)]
+pub(super) struct TryNodeContextStack(RefCell<Vec<TryNodeContext>>);
+
+impl TryNodeContextStack {
+    /// Push a new [`TryNodeContext`] for recording intermediate states
+    /// while visiting a [`ruff_python_ast::StmtTry`] node that has a `finally` branch.
+    pub(super) fn push_context(&self) {
+        self.0.borrow_mut().push(TryNodeContext::new());
+    }
+
+    /// Pop a [`TryNodeContext`] off the stack.
+    pub(super) fn pop_context(&self) {
+        self.0
+            .borrow_mut()
+            .pop()
+            .expect("Cannot pop a `try` block off an empty `TryBlockContexts` stack");
+    }
+
+    pub(super) fn borrow_mut(&self) -> TryNodeContextStackRefMut {
+        TryNodeContextStackRefMut(self.0.borrow_mut())
+    }
+}
+
+#[derive(Debug)]
+pub(super) struct TryNodeContextStackRefMut<'s>(RefMut<'s, Vec<TryNodeContext>>);
+
+impl<'s> TryNodeContextStackRefMut<'s> {
+    /// For each `try` block on the stack, push the snapshot onto the `try` block
+    pub(super) fn record_definition(&mut self, builder: &SemanticIndexBuilder) {
+        for context in self.0.iter_mut() {
+            context.record_definition(builder);
+        }
+    }
+
+    fn current_try_node(&mut self) -> &mut TryNodeContext {
+        self.0.last_mut().expect(
+            "`current_try_node()` should never be called if there are no `try` blocks on the stack",
+        )
+    }
+
+    /// Adjust internal state of the innermost `try` node on the stack
+    /// to reflect that a `try` block is being exited
+    /// in the [`ruff_python_ast::StmtTry`] node being visited.
+    ///
+    /// ## Panics
+    ///
+    /// Panics if there are no `try` blocks on the current stack.
+    pub(super) fn exit_try_block(&mut self) -> Vec<FlowSnapshot> {
+        self.current_try_node().exit_try_block()
+    }
+}
+
+/// Context for tracking definitions over the course of a single
+/// [`ruff_python_ast::StmtTry`] node
+///
+/// It will likely be necessary to add more fields to this struct in the future
+/// when we add more advanced handling of `finally` branches.
+#[derive(Debug)]
+struct TryNodeContext {
+    /// This field is `Some` when we are visiting the `try` suite of a
+    /// [`ruff_python_ast::StmtTry`] node; otherwise, it's `None`.
+    try_suite_snapshots: Option<Vec<FlowSnapshot>>,
```

Comment:
Is it important to know if it is `Some` or `None` because i dont' see that the code below makes use of the fact other than in an assertion. If not, consider simplifying this to a `Vec` and using `std::mem::take` to consume the value (or change `exit_try_block` to take `self`)

Reply:
Thanks, I think this was leftover complexity from the earlier version when I was trying (and failing) to handle `finally` blocks properly